_posts/2016-01-26-variance-and-functors.md

```diff
@@ -0,0 +1,309 @@
+    def imap[A, B](fa: Array[A])(f: A => B)(g: B => A): Array[B] =
+      new Array[B] {
+        // Convert read A to B before returning - covariance
+        override def read(i: Int): B =
+          f(fa.read(i))
+
+        // Convert B to A before writing - contravariance
+        override def write(i: Int, b: B): Unit =
+          fa.write(i, g(a))
+      }
+  }
+```
+
+## SerDe
+Another example of a read-write type that doesn't involve `Array`s (or mutation) can be
+found by just combining the `Read` and `Show` interfaces:
+
+```scala
+trait Serializer[A] extends Read[A] with Show[A] {
+  def read(s: String): Option[A]
+  def show(a: A): String
+}
+```
+
+`Serializer` both reads (from a `String`) and writes (to a `String`). We can't make it
+covariant because that would cause issues with `show`, and we can't make it contravariant
+because that would cause issues with `read`. Therefore our only choice is to keep it
+invariant.
+
+```scala
+val serializerInvariant: Invariant[Serializer] =
+  new Invariant[Serializer] {
+    def imap[A, B](fa: Serializer[A])(f: A => B)(g: B => A): Serializer[B] =
+      new Serializer[B] {
+        def read(s: String): Option[B] = fa.read(s).map(f)
+        def show(b: B): String = fa.show(g(b))
+      }
+  }
+```
+
+# Bringing it all together
+We can see the `Invariant` interface is more general than both `Functor` and `Contravariant` -
+where `Invariant` requires functions going in both directions, `Functor` and `Contravariant` only
+require one. We can make `Functor` and `Contravariant` subtypes of `Invariant` by ignoring
+the direction we don't care about.
+
+```scala
+trait Functor[F[_]] extends Invariant[F] {
+  def map[A, B](fa: F[A])(f: A => B): F[B]
+
+  def imap[A, B](fa: F[A])(f: A => B)(g: B => A): F[B] =
+    map(fa)(f)
+}
+
+trait Contravariant[F[_]] extends Invariant[F] {
+  def contramap[A, B](fa: F[A])(f: B => A): F[B]
+
+  def imap[A, B](fa: F[A])(f: A => B)(g: B => A): F[B] =
+    contramap(fa)(g)
+}
+```
+
+Going back to treating `Array` and `Serializer` as a read/write store, if we make it read-only (like a read-only
+handle on a resource) we can safely treat it as if it were covariant. If we are asked to read
+`Shape`s and we know how to read `Circle`s, we can read a `Circle` and upcast it into a `Shape`
+before handing it over.
+
+Similarly if we make it write-only (like a write-only handle on a resource) we can safely treat
+it as contravariant. If we are asked to store `Circle`s and we know how to store `Shape`s,
+we can upcast each `Circle` into a `Shape` before storing it.
+
+Variance manifests in two levels: one at the type level where subtyping relationships are defined, and
+the other at the value level where it is encoded as an interface which certain types can conform to.
+The types that can implement `Functor` are precisely those that can be made covariant, and the types that
+can implement `Contravariant` are precisely those that can be made contravariant.
```

Comment:
I would not say this; it would be easiest to show a counterexample in both directions. I'll focus on `Functor`, but the same problems I discuss apply to `Contravariant` and `Phantom` too.

## Covariant parameter, not functor: openly subclassable

```scala
trait Has[+A] {
  def it: A
}

// in some downstream module
trait Saves[A] extends Has[A] {
  var it: A
}
```

Whatever your `map` does, it can't return a `Saves`, unless you change `Has` to include a subtyping-polymorphic `map`, at which point it isn't the same type anymore. So `Functor` on `Has` can't meet the identity law. (`Seq` et al have the same problem.)

Side-effects introduce a lot more problems in this category, but they aren't necessary for it to be a problem; open subclassing is sufficient to break the relationship.

## Functor, not covariant parameter: higher-kind without variance variable

In this type,

```scala
final case class FA[F[_], A](fa: F[A])
```

the variance of `A` depends on the variance of `F`. Variance subsumption (à la SI-2066) means that we can pass covariant, contravariant, or even phantom (supposing it existed) type constructors as `F`. However, for that to carry over to the variance of `A`, you would need "variance variables" (think a specialization of kind variables; the variance of a type parameter position is an element of the kind of the type constructor surrounding it).

This is the # 2 thing missing from Scala's variance models: variance quantification.

Meanwhile, it is trivial to model this relationship with typeclasses: `Functor[F]` → `Functor[FA[F, ?]]`, `Contravariant[F]` → `Contravariant[FA[F, ?]]`, and `Phantom[F]` → `Phantom[FA[F, ?]]`